test/jdk/java/lang/Thread/crac/ThreadSleepOnCRPauseTest.java

```diff
@@ -0,0 +1,140 @@
+// Copyright 2019-2021 Azul Systems, Inc.  All Rights Reserved.
+// DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+//
+// This code is free software; you can redistribute it and/or modify it under
+// the terms of the GNU General Public License version 2 only, as published by
+// the Free Software Foundation.
+//
+// This code is distributed in the hope that it will be useful, but WITHOUT ANY
+// WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
+// A PARTICULAR PURPOSE.  See the GNU General Public License version 2 for more
+// details (a copy is included in the LICENSE file that accompanied this code).
+//
+// You should have received a copy of the GNU General Public License version 2
+// along with this work; if not, write to the Free Software Foundation, Inc.,
+// 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+//
+// Please contact Azul Systems, 385 Moffett Park Drive, Suite 115, Sunnyvale,
+// CA 94089 USA or visit www.azul.com if you need additional information or
+// have any questions.
+
+/*
+ * @test ThreadSleepOnCRPauseTest.java
+ * @requires (os.family == "linux")
+ * @library /test/lib
+ * @summary check if the Thread.sleep() will be completed on restore immediately
+ *          if its end time fell on the CRaC pause period
+ *          (i.e. between the checkpoint and restore)
+ * @run main ThreadSleepOnCRPauseTest
+ */
+
+
+
+import jdk.test.lib.process.OutputAnalyzer;
+import jdk.test.lib.process.ProcessTools;
+
+import java.util.concurrent.CountDownLatch;
+
+
+public class ThreadSleepOnCRPauseTest {
+
+    private final static long SLEEP   = 1500; // [ms]
+    private final static long CRPAUSE = 2500; // [ms]
+
+    private final static long EPS = 300_000_000; // [ns], i.e. 0.3 s
```

Comment:
I would like to understand more why these values are such. Was the intention `SLEEP == CRPAUSE - 1000`?

I'm also not sure why SLEEP and CRPAUSE are not even, is 500ms significant? Is it related to EPS == 300ms?

Reply:
a bit strange question, why not? they aren't too small as compared to EPS (to not worry about the proximity of the points in time), but, on the other hand, not too big to avoid a long run

but yes, the pause may be 2000 ms, 2500 ms, 2543 ms, 3000 ms etc :)

BTW: with respect to EPS - it was chosen arbitrary as well, and i think it should be provided a possibility to correct it during the run, if need (e.g., using a system property + default value)

Reply:
Fully support system property to adjust these values if needed.

So I assume, a half a second is not significant. Let's use some round number then? CRPAUSE =5 sec for example. It will be more clear to be an arbitrary chosen value for a reader.

300ms is indeed looks too much for an "a max difference between subsequent operations".